Add first mesa-llm model tutorial

niteshver opened the pull request:
This PR adds a new “first model” tutorial for **mesa-llm**, following the structure and level of detail of Mesa’s *Creating Your First Model* tutorial, but with content focused specifically on mesa-llm concepts.

### What this tutorial covers
- How mesa-llm integrates language-based reasoning into Mesa’s execution model
- Defining a minimal language-driven agent that remains a standard Mesa agent
- Using an LLM backend (Ollama with Llama 3) to replace rule-based decision logic
- Keeping Mesa’s scheduling and lifecycle unchanged
- Running and extending a simple language-reasoning model

### Notes
- Ollama is used as a lightweight local backend, but the design remains backend-agnostic.
- The example is intentionally minimal to focus on core mesa-llm ideas rather than infrastructure.

I’ve addressed the feedback from the earlier review by:
- aligning the structure and depth with Mesa’s first model tutorial,

I’d appreciate another review when you have time. Thanks!

colinfrisch committed:
link to readthedocs and small fixes

sujalgawas committed:
inital commit model.py and agent.py

sujalgawas committed:
runs with errors

sujalgawas committed:
runs with errors

sujalgawas committed:
fixed harvesting logic

sujalgawas committed:
final tests with gemini-2.5-flash-lite

sujalgawas committed:
Readme.md file added

sujalgawas committed:
readme.md

sujalgawas committed:
Fix harvesting logic in tools.py

sujalgawas committed:
review changes

niteshver committed:
created agent and model class

niteshver committed:
create agent and model with ollama

niteshver committed:
correct terminal

niteshver committed:
some additional changes

niteshver committed:
mesa_llm tutorial is now done

sujalgawas committed:
review 2

niteshver committed:
Improve reasoning demonstration in introductory model tutorial

niteshver commented:
Hi @colinfrisch, I’ve implemented a reasoning-focused version of the model where agents explicitly explain their reasoning before deciding on an action, and updated the tutorial to reflect this behavior. Happy to make further adjustments if needed.

februarysea committed:
fix: treat ollama_chat as ollama provider (#43)

colinfrisch commented on docs/first_model.md:
```diff
@@ -0,0 +1,248 @@
+* Ollama
+* OpenRouter
+* NovitaAI
+* Gemini
```
Not necessary here, supported models are already in the docs

colinfrisch commented on docs/first_model.md:
````diff
@@ -0,0 +1,248 @@
+```python
+import mesa_llm
+import mesa
+import ollama
````
We will add ollama dependency directly in mesa-llm. No need to put it here

colinfrisch reviewed (changes requested):
Essentially, you have not used any code from the mesa-llm library here, please try to build a working model yourself to check out how this library works before contributing as this is supposed to be a tutorial for mesa-llm. Also please be careful what tools you use. An excessive use of AI can result in a temporary ban for the mesa ecosystem and have definitive consequences on GSoC.

niteshver commented:
Hi @colinfrisch,

I’ve rebuilt the First Model using LLMAgent and ReActReasoning, keeping it minimal and aligned with the examples.

Before moving forward, I just want to confirm scope: for the First Model tutorial, is the expectation to keep the example very simple (e.g. a single agent focused on reasoning), avoid printing or exposing reasoning output, and leave backends like Ollama for later or more advanced examples?

I want to make sure I’m fully aligned before refining the documentation further.

niteshver commented:
Hi @colinfrisch,

I rebuilt the First Model to focus strictly on using LLMAgent with a Reasoning , keeping everything else minimal. My goal is to demonstrate how LLM-based reasoning is integrated into a Mesa model, without adding extra behavior yet.

Just wanted to confirm that this aligns with the intended scope of the tutorial.

colinfrisch committed:
Fix and add dependencies (#49)

* fix-prompt-gestion

* fix-prompt-gestion align testing

* fix dependencies

niteshver committed:
finalize first mesa-llm model tutorial focused on LLMAgent

colinfrisch committed:
Fix prompt gestion (#48)

* fix-prompt-gestion

* fix-prompt-gestion align testing

dependabot[bot] committed:
Bump actions/upload-artifact from 5 to 6 (#50)

Bumps [actions/upload-artifact](https://github.com/actions/upload-artifact) from 5 to 6.
- [Release notes](https://github.com/actions/upload-artifact/releases)
- [Commits](https://github.com/actions/upload-artifact/compare/v5...v6)

---
updated-dependencies:
- dependency-name: actions/upload-artifact
  dependency-version: '6'
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump actions/cache from 4 to 5 (#51)

Bumps [actions/cache](https://github.com/actions/cache) from 4 to 5.
- [Release notes](https://github.com/actions/cache/releases)
- [Changelog](https://github.com/actions/cache/blob/main/RELEASES.md)
- [Commits](https://github.com/actions/cache/compare/v4...v5)

---
updated-dependencies:
- dependency-name: actions/cache
  dependency-version: '5'
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

niteshver commented:
Thank you for the feedback. I’ll go with a negotiation model, since I’m already familiar with this setup from the existing example and feel confident using it to clearly demonstrate how LLM agents reason and interact. I’ll keep the model small and tutorial-friendly while focusing on clarity and purpose

februarysea committed:
test: update module_llm tests for system prompt handling (None -> "") (#44)

sujalgawas committed:
test disabling Codecov upload

niteshver committed:
Add tutorial-friendly negotiation model

niteshver committed:
Minor cleanup

niteshver committed:
Delete accidently created file

sujalgawas committed:
test build

sujalgawas committed:
testing id-token: write

sujalgawas committed:
reset

colinfrisch reviewed (changes requested):
It's starting to look like something we could merge ! Little suggestions down below. Also, for the negotiation tutorial, I think that it would be the right place to demonstrate the messaging system between agents : if you can manage to make use of it and make  it tutorial friendly, it would be great !

Last thing also : when you write code in your tutorials, don't hesitate to add a lot of comments to explain directly in the code the role of each function/method/attribute, etc.

niteshver commented on docs/first_model.md:
```diff
@@ -0,0 +1,222 @@
+- In this introductory tutorial, action suggestions are not executed.
+- Actions are shown only as part of the reasoning trace.
+- Environments and action execution are introduced in later tutorials.
+
```
thanks! I’ll add a small exercise section at the end of the tutorial, similar to Mesa’s first model tutorial.

colinfrisch committed:
Update docs/negotiation_model_tutorial.md

niteshver committed:
Add agent messaging and documentation fixes to negotiation tutor

niteshver committed:
fix first_model & negotiation_model_tutorial

niteshver committed:
fix first model

Nithin9585 committed:
feat: Add dialogue history context to negotiation agents
- Implement _get_dialogue_history() to extract conversation from memory
- Inject dialogue context into reasoning prompts
- Support both STLTMemory and EpisodicMemory
- Handle both agent objects (send_message) and IDs (speak_to tool)
- Format senders as 'AgentType ID' for clarity
- Limit to last 5 exchanges to manage context window

Fixes agent amnesia in multi-turn negotiations where agents
forgot previous offers and repeated themselves.

Nithin9585 committed:
refactor: address code review feedback - remove duplication, optimize iteration, remove debug prints

niteshver commented:
Hi @colinfrisch 

I’m planning to work on a new mesa-llm example focused on opinion dynamics, where agents use language models to communicate while opinion updates.

The model would include:
1. Agents holding a numeric opinion on a single topic
2. LLM-generated messages exchanged between neighboring agents
3. Explicit trust updates (increase on agreement, decrease on disagreement)
4. A graph-based visualization showing opinion and trust evolution over time

Please let me know if this direction sounds appropriate before I start implementing it.

colinfrisch reviewed (approved):
Looks good to me !! @sanika-n what do you think about this tutorial (especially the negociation part)

niteshver commented:
Hi @colinfrisch 
Just checking if any further changes are needed in this PR from my side

sanika-n reviewed (changes requested):
Hey, it looks great to me, thanks a lot for working on it. Just mentioned some really quick changes, after that's done, I think we will be good to merge:) So sorry for the delay...

niteshver committed:
docs: minor fixes to first model and negotiation_model

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci
